### PR TITLE
Support PSNR algorithm in checkbox-support (New)

### PR DESCRIPTION
**Pull request by baconYao**
## Description

Implement the PSNR algorithm in Checkbox Support by reference the OpenCV official document.
  - https://docs.opencv.org/3.4/d5/dc4/tutorial_video_input_psnr_ssim.html

Why I propose this PR?

- There will have a bunch of Video Codec Scenarios that rely on PSNR comparison, therefore, I reference the PSNR implementation of OpenCV.

## Resolved issues

N/A

## Documentation

OpenCV official document:

- https://docs.opencv.org/3.4/d5/dc4/tutorial_video_input_psnr_ssim.html

### CODE REVIEW

**Comment by baconYao** [comment]
Hi @kissiel and @fernando79513, I noticed that unittest cannot find the module `numpy`, may I know how can I solve it?

```bash
==================================== ERRORS ====================================
_____________ ERROR collecting checkbox_support/tests/test_psnr.py _____________
ImportError while importing test module '/home/runner/work/checkbox/checkbox/checkbox-support/checkbox_support/tests/test_psnr.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
checkbox_support/tests/test_psnr.py:2: in <module>
    import numpy as np
E   ModuleNotFoundError: No module named 'numpy'
```

**Comment by fernando79513** [comment]
> Hi @kissiel and @fernando79513, I noticed that unittest cannot find the module `numpy`, may I know how can I solve it?
> 
> ```shell
> ==================================== ERRORS ====================================
> _____________ ERROR collecting checkbox_support/tests/test_psnr.py _____________
> ImportError while importing test module '/home/runner/work/checkbox/checkbox/checkbox-support/checkbox_support/tests/test_psnr.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/importlib/__init__.py:127: in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
> checkbox_support/tests/test_psnr.py:2: in <module>
>     import numpy as np
> E   ModuleNotFoundError: No module named 'numpy'
> ```

At the moment the numpy module is not installed for checkbox support tests. On the base provider, we are installing opencv in the github runners. This can be done including the dependency on the `tox.ini` files.

Regarding if we should include numpy in checkbox-support, its a great point to be made. As @kissiel mentioned, numpy is quite a big library. It takes about 100MB of space, so maybe we should try to avoid including it in checkbox-support

File: checkbox-support/checkbox_support/tests/test_psnr.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+    @patch("checkbox_support.scripts.psnr._get_frame_resolution")
+    @patch("checkbox_support.scripts.psnr.cv2.VideoCapture")
+    def test_get_average_psnr_different_dimensions(
+        self, mock_videocapture, mock_get_frame_resolution
+    ):
+        mock_capture_ref = mock_videocapture.return_value
+        mock_capture_test = mock_videocapture.return_value
+        mock_capture_test.isOpened.return_value = True
+        mock_capture_ref.isOpened.return_value = True
```
Comment (fernando79513) [inline checkbox-support/checkbox_support/tests/test_psnr.py:106]: These two are the same mock, you could use just one:
```suggestion
    def test_get_average_psnr_different_dimensions(
        self, mock_vc, mock_get_frame_resolution
    ):
        mock_vc.return_value.isOpened.return_value = True
```

File: checkbox-support/checkbox_support/tests/test_psnr.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            get_average_psnr("ref_file.mp4", "test_file.mp4")
+
+    @patch("checkbox_support.scripts.psnr.cv2.VideoCapture")
+    @patch("checkbox_support.scripts.psnr._get_psnr")
+    @patch("checkbox_support.scripts.psnr._get_frame_resolution")
+    def test_get_average_psnr(
+        self, mock_get_frame_resolution, mock_get_psnr, mock_VideoCapture
+    ):
+        # Create a mock VideoCapture object
+        mock_video_capture = MagicMock()
```
Comment (fernando79513) [inline checkbox-support/checkbox_support/tests/test_psnr.py:119]: Same as before

**Review by fernando79513 — COMMENTED** [review]
Sorry for the late review. I didn't see any big issues, just some small improvements.
The only thing that we should take into account is that usually, a low PSNR means bad quality.
If we are assigning 0 to the case when there is no loss, we should also add a check for 0 when we assign the threshold on the final test.

File: checkbox-support/pyproject.toml [hunk 3]
```diff
@@ -40,6 +40,7 @@
   checkbox-support-parse = "checkbox_support.parsers:main"
   checkbox-support-zapper-proxy = "checkbox_support.scripts.zapper_proxy:main"
   checkbox-support-image_checker = "checkbox_support.scripts.image_checker:main"
+  checkbox-support-psnr = "checkbox_support.scripts.psnr:main"
```
[the remark this answers is not in the document]
Comment (baconYao) [reply checkbox-support/pyproject.toml:43]: Indeed, there's no requirement to expose it now. I could remove it and let us import it if used needs.

[the remark this answers is not in the document]
Comment (baconYao) [reply checkbox-support/pyproject.toml]: Removed

File: checkbox-support/checkbox_support/tests/test_psnr.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            test_file="test.mp4",
+            show_psnr_each_frame=False,
+        )
+        args = psnr_args().parse_args()
+
+        self.assertEqual(args.reference_file, "ref.mp4")
+        self.assertEqual(args.test_file, "test.mp4")
+        self.assertFalse(args.show_psnr_each_frame)
+
+    @patch("checkbox_support.scripts.psnr.argparse.ArgumentParser.parse_args")
```
[the remark this answers is not in the document]
Comment (baconYao) [reply checkbox-support/checkbox_support/tests/test_psnr.py]: Fix

File: checkbox-support/checkbox_support/tests/test_psnr.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    def test_identical_images(self):
+        img1 = self.create_image(100, 100, 255)
+        img2 = self.create_image(100, 100, 255)
+        self.assertEqual(_get_psnr(img1, img2), 0.0)
+
+    def test_different_images(self):
+        img1 = self.create_image(100, 100, 255)
+        img2 = self.create_image(100, 100, 0)
+        self.assertNotEqual(_get_psnr(img1, img2), 0.0)
+
```
[the remark this answers is not in the document]
Comment (baconYao) [reply checkbox-support/checkbox_support/tests/test_psnr.py:70]: Thanks @fernando79513. I misunderstood your meaning last commit, therefore, I though I did it then resolved this thread.

Yes, your suggestion is great, let me commit it.

**Review by fernando79513 — APPROVED** [review]
LGTM +1!

**Pull request merged**